lib/ansible/module_utils/tacp_ansible/tacp_utils.py

```diff
@@ -179,64 +182,215 @@ def power_action_on_instance_by_uuid(self, uuid, power_action):
             Action.ABSENT: self.api.delete_application_using_delete,
             Action.RESUMED: self.api.resume_application_using_put
         }
-        return power_action_dict[power_action](uuid)
-
-
-class ApplicationGroupResource(Resource):
-    resource_class = tacp.ApplicationGroupsApi
-
-    def filter(self, **filters):
-        return self.api.get_application_group_list_using_get(
-            **self.get_filters_kws(**filters)
-        )
-
-    @wait_to_complete
-    def create(self, name, datacenter_uuid):
-        return self.api.create_application_group_using_post(
-            tacp.ApiCreateApplicationGroupPayload(
-                name=name, datacenter_uuid=datacenter_uuid
-            )
-        )
+        return power_action_dict[power_action](uuid).to_dict()
 
 
 class VlanResource(Resource):
     resource_class = tacp.VlansApi
 
-    def filter(self, **filters):
-        return self.api.get_vlans_using_get(
-            **self.get_filters_kws(**filters)
-        )
+    def filter(self, as_dict=False, **filters):
```

Comment:
This is the type of thing that I wanted the parent `Resource` class to take care of and not repeat on all the subclasses. If you have the subclasses implement their specific implementations as "private" methods:
```python
def _filter(self, **filters):
    return self.api_get_vlans_using_get(**filters)
```

The parent class can take care of the rest in the publicly exposed methods:
```python
def filter(self, as_dict=Flase, **filters):
    filters = self.get_filters(**filters)
    results = self._filter(**filters)
    if as_dict:
        return [result.to_dict() for result in
                    results]
    return results
```

I know it is strange to implement private methods in the subclasses but it is a cleaner DRY-er approach, in my opinion.
@mariusvigariu inputs?

Reply:
I believe the best way is to deal with tacp payload instances, rather than dicts. When I suggested `as_dict` @xmadsen that was like a theoretically alternative, but it does not really suit here as you can see.

I suggest to remove the `to_dict` processing in here and in the rest of the resources code and have all methods return tacp payload instances. Let the callers process the instances to dict or however they see fit.

@patrickjennings it is cleaner indeed, i would just castle the methods have the children implement filter and work on the base private filter. But even better I would suggest to just have a filter endpoint class attribute which can be changed by resource subclasses, because it seems like each api class has a dedicated method of doing GETs for items.
```

class Resource(object):                                                         
    # ...                                                                       
    filter_method = None                                                        
                                                                                
    def filter(self, **filters):                                                
        if self.filter_method is None:                                          
            raise Exception('Invalid self.filter_method')                       
        method = getattr(self.api, self.filter_method, None)                    
        if method is None:                                                      
            raise Exception('Invalid self.filter_method')                       
                                                                                
        return method(**self.get_filters__kws(**filters))                       
                                                                                
                                                                                
class SomeResource(object):                                                     
    # resource_class = ...                                                      
    filter_method = 'get_applications_using_get'  

```